packages/flutter_rx_bloc/lib/src/base/stream_multi_builder_3.dart

```diff
@@ -0,0 +1,266 @@
+  _StreamContainer({
+    required super.stream,
+    this.initialData,
+  });
+
+  final T? initialData;
+
+  @override
+  AsyncSnapshot<T> initial() => initialData == null
+      ? AsyncSnapshot<T>.nothing()
+      : AsyncSnapshot<T>.withData(ConnectionState.none, initialData as T);
+
+  @override
+  AsyncSnapshot<T> afterConnected(AsyncSnapshot<T> current) =>
+      current.inState(ConnectionState.waiting);
+
+  @override
+  AsyncSnapshot<T> afterData(AsyncSnapshot<T> current, T data) =>
+      AsyncSnapshot<T>.withData(ConnectionState.active, data);
+
+  @override
+  AsyncSnapshot<T> afterError(
+          AsyncSnapshot<T> current, Object error, StackTrace stackTrace) =>
+      AsyncSnapshot<T>.withError(ConnectionState.active, error, stackTrace);
+
+  @override
+  AsyncSnapshot<T> afterDone(AsyncSnapshot<T> current) =>
+      current.inState(ConnectionState.done);
+
+  @override
+  AsyncSnapshot<T> afterDisconnected(AsyncSnapshot<T> current) =>
+      current.inState(ConnectionState.none);
+}
+
+/// endregion
+
+/// region _StreamContainerBase
+
+abstract class _StreamContainerBase<T, S> {
+  const _StreamContainerBase({required this.stream});
+
+  final Stream<T>? stream;
+
+  S initial();
+
+  S afterConnected(S current);
+
+  S afterData(S current, T data);
+
+  S afterError(S current, Object error, StackTrace stackTrace);
+
+  S afterDone(S current);
+
+  S afterDisconnected(S current);
+}
+
+/// endregion
+
+/// region _StreamSubscriptionContainer
+
+class _StreamSubscriptionContainer<T, S> {
+  StreamSubscription<T>? subscription; // ignore: cancel_subscriptions
+  late S summary;
+
+  void update(
+    Stream<T>? oldStream,
+    _StreamContainerBase<T, S> streamContainer,
+    void Function(void Function()) setState,
+  ) {
+    if (oldStream != streamContainer.stream) {
+      if (subscription != null) {
+        unsubscribe();
+        summary = streamContainer.afterDisconnected(summary);
+      }
+      subscribe(streamContainer, setState);
+    }
+  }
+
+  void subscribe(
+    _StreamContainerBase<T, S> streamContainer,
+    _StateUpdateFunction setState,
+  ) {
+    if (streamContainer.stream != null) {
+      subscription = streamContainer.stream!.listen((T data) {
+        setState(() {
+          summary = streamContainer.afterData(summary, data);
+        });
+      }, onError: (Object error, StackTrace stackTrace) {
+        setState(() {
+          summary = streamContainer.afterError(summary, error, stackTrace);
+        });
+      }, onDone: () {
+        setState(() {
+          summary = streamContainer.afterDone(summary);
+        });
+      });
+      summary = streamContainer.afterConnected(summary);
+    }
+  }
+
+  void unsubscribe() {
+    if (subscription != null) {
+      subscription!.cancel();
+      subscription = null;
+    }
+  }
+}
+
+/// endregion
+
+/// region _AsyncWidgetBuilder and _StateUpdateFunction
+
+typedef _AsyncWidgetBuilder<T1, T2, T3> = Widget Function(
```

Comment:
I also tested with this lint rule : [library_private_types_in_public_api](https://dart.dev/tools/linter-rules/library_private_types_in_public_api) and it suggests that  `_AsyncWidgetBuilder` should be made not private